### ISSUE TITLE
Test failure in 1.7.5

### ISSUE THREAD

**Issue by lfam** [lz4/lz4#308]
Greetings!

I'm working on distro packaging for GNU Guix [0]. I'm trying to update our lz4 package to 1.7.5, but the test suite is failing like this:

```
make -C tests test
make[1]: Entering directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/tests'
make -C ../programs clean lz4 CFLAGS="-O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1 "
make[2]: Entering directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/programs'
Cleaning completed
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o ../lib/lz4.o ../lib/lz4.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o ../lib/lz4hc.o ../lib/lz4hc.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o ../lib/xxhash.o ../lib/xxhash.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o ../lib/lz4frame.o ../lib/lz4frame.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o datagen.o datagen.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o lz4io.o lz4io.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o lz4cli.o lz4cli.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  -c -o bench.o bench.c
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -DXXH_NAMESPACE=LZ4_  ../lib/lz4.o ../lib/lz4hc.o ../lib/xxhash.o ../lib/lz4frame.o datagen.o lz4io.o lz4cli.o bench.o -o lz4
make[2]: Leaving directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/programs'
gcc -O3  -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_  -I../programs ../programs/datagen.c datagencli.c -o datagen
\n ---- test opt-parser ----
./datagen -g16KB      | ../programs/lz4 -12      | ../programs/lz4 -t

stdin                : decoded 16384 bytes                                     
./datagen -P10        | ../programs/lz4 -12B4    | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 65536 bytes                                     
./datagen -g256K      | ../programs/lz4 -12B4D   | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 262144 bytes                                    
./datagen -g512K -P25 | ../programs/lz4 -12BD    | ../programs/lz4 -t

stdin                : decoded 524288 bytes                                    
./datagen -g1M        | ../programs/lz4 -12B5    | ../programs/lz4 -t
using blocks of size 256 KB 

stdin                : decoded 1048576 bytes                                   
./datagen -g2M -P99   | ../programs/lz4 -11B4D   | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 2097152 bytes                                   
./datagen -g4M        | ../programs/lz4 -11vq    | ../programs/lz4 -qt

./datagen -g8M        | ../programs/lz4 -11B4    | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 8388608 bytes                                   
./datagen -g16M -P90  | ../programs/lz4 -11B5    | ../programs/lz4 -t
using blocks of size 256 KB 

stdin                : decoded 16777216 bytes                                  
./datagen -g32M -P10  | ../programs/lz4 -11B5D   | ../programs/lz4 -t
using blocks of size 256 KB 

stdin                : decoded 33554432 bytes                                  
make[2]: Entering directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/programs'
ln -s lz4 unlz4
make[2]: Leaving directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/programs'
make[2]: Entering directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/programs'
ln -s lz4 lz4cat
make[2]: Leaving directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/programs'
\n ---- test lz4 basic compression/decompression ----
./datagen -g0     | ../programs/lz4 -v     | ../programs/lz4 -t
*** LZ4 command line interface 64-bits v1.7.5, by Yann Collet ***

Compressed 0 bytes into 15 bytes ==> 1500.00%                                  
stdin                : decoded 0 bytes                                         
./datagen -g16KB  | ../programs/lz4 -9     | ../programs/lz4 -t

stdin                : decoded 16384 bytes                                     
./datagen -g20KB > tmpSrc

../programs/lz4 < tmpSrc   | ../programs/lz4 -d > tmpRes
diff -q tmpSrc tmpRes
../programs/lz4 --no-frame-crc < tmpSrc | ../programs/lz4 -d > tmpRes
diff -q tmpSrc tmpRes
./datagen         | ../programs/lz4        | ../programs/lz4 -t

stdin                : decoded 65536 bytes                                     
./datagen -g6M -P99 | ../programs/lz4 -9BD | ../programs/lz4 -t

stdin                : decoded 6291456 bytes                                   
./datagen -g17M   | ../programs/lz4 -9v    | ../programs/lz4 -qt
*** LZ4 command line interface 64-bits v1.7.5, by Yann Collet ***

Compressed 17825792 bytes into 8086251 bytes ==> 45.36%                        
./datagen -g33M   | ../programs/lz4 --no-frame-crc | ../programs/lz4 -t

stdin                : decoded 34603008 bytes                                  
./datagen -g256MB | ../programs/lz4 -vqB4D | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 268435456 bytes                                 
../programs/lz4 --rm -f tmp
"Md@?
     ?hello world
??R?ls -ls tmp         && false || true   # must fail (--rm)
ls: cannot access 'tmp': No such file or directory
ls -ls tmp.lz4
ls: cannot access 'tmp.lz4': No such file or directory
make[1]: *** [Makefile:238: test-lz4-basic] Error 2
make[1]: Leaving directory '/tmp/guix-build-lz4-1.7.5.drv-0/lz4-1.7.5/tests'
make: *** [Makefile:105: test] Error 2
```

The strange corruption that begins with ```"Md@?``` is reproducible; every time I run the test suite, it fails like that.

I'm building on Linux 4.9, using the ext4 filesystem. Please let me know what other information I can provide.

[0]
http://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/compression.scm#n603

**Comment by Cyan4973** [comment]
Thanks for your detailed report @lfam .

My current understanding is that the shell executing the test does not interpret as expected the following command : 
`ls -ls tmp && false || true`

The test checks if file `tmp` is still there.
As stated in the comments, the `ls` command is _supposed_ to fail, proving that `tmp` is no longer there.
The failure is converted in to a success, and vice-versa, with `&& false || true`.

So apparently, that part doesn't work on your test system.
It would be interesting to know why, and to find a work-around.

Unfortunately, I'm currently not able to reproduce the problem, as the test pass as intended on all my available tests systems and VM. So I probably need another one to reproduce your result.

**Comment by Cyan4973** [comment]
My previous answer supposed that the error came from : 
`ls -ls tmp && false || true`.

Now, I believe that wasn't correct.
The problem comes instead from the second `ls` command : 
`ls -ls tmp.lz4`

It's quite different.
This command fails because `tmp.lz4` does not exist. But it should.

It was supposed to be created by : `../programs/lz4 --rm -f tmp`.
In above command `tmp.lz4` is the implicit destination name that should be created.

But apparently, the result is instead pushed to `stdout`.
That's why we can see the output : 
` "Md@?`
`    ?hello world`


I suspect a problem of specification.
`man lz4` tells : 
>        When no destination name is provided, compressed file name receives a .lz4 suffix  
>        When no destination name is provided, if stdout is not the console, it becomes the output (like a silent -c)

The second line seems to explain the issue :

when the test is run from the console, the first line applies : `compressed file name receives a .lz4 suffix `

But maybe your test is run from _something which is not detected as a console_ . It's not clear what it is (it depends on your test chain). But let's assume `stdout` is not considered to be a console, then the second line applies, and now result is pushed to `stdout`.

It seems the specification itself is problematic, and would need a change.
But the change could also break existing scripts.
So it's pretty tricky ....

A quickfix could be to simply avoid this situation, for example by rewriting the failing test as : `../programs/lz4 --rm -f tmp tmp.lz4`
in which case, the destination file is explicit, so it doesn't rely on above rule.

But that wouldn't change the underlying issue ...

**Comment by mwgamera** [comment]
If the test requires stdout to be a terminal (you are checking if it's a terminal) then maybe it should explicitly make it a terminal. I was able to reproduce the problem with `make -C tests test | cat` and the following appears to fix it (I'm not sure how portable it is, though):
```diff
diff --git a/tests/Makefile b/tests/Makefile
index 5f8efb7..a0d3764 100644
--- a/tests/Makefile
+++ b/tests/Makefile
@@ -235,19 +235,19 @@ test-lz4-basic: lz4 datagen unlz4 lz4cat
        ./datagen -g33M   | $(LZ4) --no-frame-crc | $(LZ4) -t
        ./datagen -g256MB | $(LZ4) -vqB4D | $(LZ4) -t
        @echo "hello world" > tmp
-       $(LZ4) --rm -f tmp
+       $(LZ4) --rm -f tmp > /dev/tty
        ls -ls tmp         && false || true   # must fail (--rm)
        ls -ls tmp.lz4
        $(PRGDIR)/lz4cat tmp.lz4              # must display hello world
        ls -ls tmp.lz4
-       $(PRGDIR)/unlz4 --rm tmp.lz4
+       $(PRGDIR)/unlz4 --rm tmp.lz4 > /dev/tty
        ls -ls tmp
        ls -ls tmp.lz4     && false || true   # must fail (--rm)
        ls -ls tmp.lz4.lz4 && false || true   # must fail (unlz4)
        $(PRGDIR)/lz4cat tmp                  # pass-through mode
        ls -ls tmp
        ls -ls tmp.lz4     && false || true   # must fail (lz4cat)
-       $(LZ4) tmp                         # creates tmp.lz4
+       $(LZ4) tmp > /dev/tty              # creates tmp.lz4
        $(PRGDIR)/lz4cat < tmp.lz4 > tmp3  # checks lz4cat works with stdin (#285)
        $(DIFF) -q tmp tmp3
        $(PRGDIR)/lz4cat < tmp > tmp2      # checks lz4cat works with stdin (#285)
```

By the way, `whatever && false || true` is *always* true. It won't fail when the file is removed as requested, but it won't either when it is not. You want to use `! whatever`. And on a side note, I think these should use test rather than ls.

**Comment by Cyan4973** [comment]
These are very good points @mwgamera .
I'll update the `Makefile` to follow your recommendations.

**Comment by Cyan4973** [comment]
Tests were fixed thanks to patches from @ldv-alt , @iburinoc  and @esiegerman

**Comment by lfam** [comment]
Thanks to everyone for the analysis and fixes!

I cherry-picked the fixes and lz4 1.7.5 is working fine:

https://hydra.gnu.org/job/gnu/master/lz4-1.7.5.x86_64-linux
